File: languages/cpp/01-foundations/functions/example/main.cpp
// Module focus: Breaking behavior into reusable functions with clear inputs and outputs.
// Why it matters: the example makes it possible to decompose a problem into focused functions
// with explicit contracts before the learner tackles the exercises.
// about.

#include <iostream>
#include <vector>
using namespace std;

// Separate helpers keep the main path focused on how to decompose a problem into focused
// functions with explicit contracts.
int sum(int a, int b) { return a + b; }

void swapByReference(int& left, int& right) {
    const int temp = left;
    left = right;
    right = temp;
}

void printVector(const vector<int>& values) {
    cout << "[";
    for (size_t i = 0; i < values.size(); ++i) {
        cout << values[i];
        if (i + 1 < values.size()) {
            cout << ", ";
        }
    }
    cout << "]\n";
}

// Fixed inputs make the consequence of unnecessary copies of large objects visible and
// repeatable.
int main() {
    // These values exercise the normal path before the exercises vary the documented boundaries.
    // The printed result shows whether the program can use parameters and return values without
    // hidden state changes.
    cout << "sum(4, 6) = " << sum(4, 6) << '\n';

    int first = 10;
    int second = 20;
    cout << "Before swap: first=" << first << ", second=" << second << '\n';
    swapByReference(first, second);
    cout << "After swap:  first=" << first << ", second=" << second << '\n';

    const vector<int> numbers{1, 2, 3, 4, 5};
    cout << "Vector content: ";
    printVector(numbers);

    return 0;
}

File: languages/python/01-foundations/functions/example/main.py
# Module focus: Breaking behavior into reusable functions with clear inputs and outputs.
# Why it matters: the example makes it possible to decompose a problem into focused functions with
# explicit contracts before the learner tackles the exercises.
# about.

# Separate helpers keep the main path focused on how to decompose a problem into focused functions
# with explicit contracts.
def add(a, b):
    return a + b


def swap_in_list(values, i, j):
    values[i], values[j] = values[j], values[i]


def print_list(values):
    print(values)


# Fixed inputs make the consequence of embedding all logic in main instead of reusable helpers
# visible and repeatable.
# The printed result shows whether the program can use parameters and return values without hidden
# state changes.
print(add(4, 6))

numbers = [10, 20, 30]
print_list(numbers)
swap_in_list(numbers, 0, 1)
print_list(numbers)
